### ISSUE TITLE
Mirror locations failing

### ISSUE THREAD

**Issue by sparklyballs** [linuxserver/docker-baseimage#10]
It appears that not all mirror sites are equal, seeing this a few times now, with sabnzbd, but i don't think it's something that would be particular to that one container, hence raising the issue here.

http://lime-technology.com/forum/index.php?topic=42441.msg426452#msg426452

E: Method gave invalid 103 Redirect message

one of the mirror repos is throwing that error and preventing update of SAB, consequently people are stuck on the version that is in the image.

**Comment by lonix** [comment]
I am looking into getting a fix for when apt mirrors are not behaving. I'll give you some more feedback tomorrow

**Comment by aptalca** [comment]
Just an update, I tried again, this time got a different mirror, but the same error:

```
root@aa4ad93732b5:/# apt-get update
Hit http://ppa.launchpad.net trusty InRelease
Ign http://mirrors.rit.edu/ubuntu/ trusty InRelease
Hit http://ppa.launchpad.net trusty/main amd64 Packages
Hit http://mirrors.rit.edu/ubuntu/ trusty-security InRelease
Hit http://mirrors.rit.edu/ubuntu/ trusty-updates InRelease
Get:1 http://mirrors.rit.edu/ubuntu/ trusty Release.gpg [933 B]
Hit http://mirrors.rit.edu/ubuntu/ trusty-security/main Sources
E: Method gave invalid 103 Redirect message
```

EDIT: I just realized sparkly pushed a new build. The above message was prior to the update. With the new build, apt-get update seems to work, but it was already up to date.

**Comment by sparklyballs** [comment]
for the mirror issue i'm thinking some kind of sed wizadry to comment/uncomment the lines of our sources.list, by doing that we preserve any additions if we're juidicial with the sed, then copy in a backup list to /etc/apt/sources.list.d. that way we can minimise file copy errors if the script is halted etc... 
something like this... (haven't quite figured the sed yet)

```
#!/bin/bash

# test conditions for aptlist and base aptilist
[ "$BASE_APTLIST" ] && APTLIST="$BASE_APTLIST ""$APTLIST"
[ -z "$APTLIST" ] && exit 0

# make sure mirror and apt lists are in proper order
[[ -f /etc/apt/sources.list.d/backup_sources.list ]] && rm /etc/apt/sources.list.d/backup_sources.list

echo "We are now refreshing packages from apt repositorys, this *may* take a while"

# try apt-get update and output exit code to variable
apt-get clean
apt-get update -qq 
RETVAL=$?

# if update with mirrors failed, try backup list
if [ "$RETVAL" -gt "0" ]; then
cp /defaults/backup_sources.list /etc/apt/sources.list.d/backup_sources.list 
echo "put the mirror replacing stuff here"
apt-get clean
apt-get update -qq
echo "put the mirror stuff back here"
fi

apt-get --only-upgrade install -yqq $APTLIST
```

**Comment by aptalca** [comment]
A solution is suggested over here: https://lists.ubuntu.com/archives/foundations-bugs/2015-August/248569.html  and here: https://kamaradski.com/2780/apt-get-update-103-redirect-hash-sum-mismatch-errors

They propose the following:

`I have learned that workaround for this issue is `cd
  /var/lib/apt/lists;rm *;cd partial;rm *;cd`.`

I didn't get a chance to test it because I already temporarily fixed my issue and it is not happening at the moment.

**Comment by sparklyballs** [comment]
that fix is very specific for "103" 

the fix here is a attempt at a catchall for mirror issues in general rather than so specific for one error.

**Comment by sparklyballs** [comment]
https://gist.github.com/sparklyballs/9d73b4712b022fd31708

**Comment by lonix** [comment]
I Think @aptalca  has a point here, also your script seems to be over doing it.

**Comment by lonix** [comment]
This issue is being handled by the failoverapt branch. from SHA: 65af9867d9e5e19f270deec0c2e36cd16e7ff6b4

**Comment by sparklyballs** [comment]
take a look at my fork here.

https://github.com/sparklyballs/docker-baseimage

i put the contents of the original (the one in the phusion base) sources.file back into the one with the mirrors that we use, modifying them with multiverse etc.

this helps mitigate package dependencies when a new package is released and hasn't propagated to the mirrors yet. the main bulk of the pull still comes from the mirror, but the package that might have failed the update "should" now get fetched from the standard ubuntu repo.

the update script does an initial update check with the mirrors in place, if it returns anything other than 0 , then it seds out the mirrors, cleans the caches and tries again, and then goes on to the upgrade phase.

**Comment by pecigonzalo** [comment]
https://bugs.launchpad.net/ubuntu/+source/apt/+bug/1479045

**Comment by sparklyballs** [comment]
some further reading on the subject, i've gleaned some insight into apt-get and how sources are handled that seem relevant to ongoing work on this issue.

apt-get prioritises repositories in sources.list in order from top to bottom within the same list and in order (not sure if alphabetical for multiple lists , it just said order) in sources.list.d

if the same version of a package or a dependency is in multiple repositories it pulls from the repo higher in the list, but if a later version is available in any other repository the later version takes priority.

that's why with mirrors and original sources in the list/lists, missing packages/dependencies from mirrors that aren't kept up to date are pulled from the original source repos.

**Comment by sparklyballs** [comment]
closed with commit 7efd63b09ba8030e64f262c03c3036188c6a347d
